query: binary encoder for repo branches

keegancsmith opened the pull request:
We implement a custom binary marshaller for a list of repos to
branches. When profiling Sourcegraph this is one of the dominant items
and this provides a significant improvement.

Note: This PR does not provide a way to roll this change out. I
investigated if this was possible with Gob and its pretty
difficult. This means that while Sourcegraph and Zoekt are out of sync
searches will fail. I am unsure if its worth making this smooth to
rollout.

```
name                   old time/op    new time/op    delta
RepoBranches_Encode-8    2.37ms ± 3%    0.62ms ± 0%   -73.77%  (p=0.000 n=10+8)
RepoBranches_Decode-8    4.19ms ± 2%    0.74ms ± 1%   -82.37%  (p=0.000 n=10+9)

name                   old bytes      new bytes      delta
RepoBranches_Encode-8     393kB ± 0%     344kB ± 0%   -12.48%  (p=0.000 n=10+10)

name                   old alloc/op   new alloc/op   delta
RepoBranches_Encode-8     726kB ± 0%     344kB ± 0%   -52.60%  (p=0.000 n=10+9)
RepoBranches_Decode-8    2.31MB ± 0%    1.44MB ± 0%   -37.51%  (p=0.000 n=9+10)

name                   old allocs/op  new allocs/op  delta
RepoBranches_Encode-8     20.0k ± 0%      0.0k ± 0%  -100.00%  (p=0.000 n=10+10)
RepoBranches_Decode-8     50.6k ± 0%      0.4k ± 0%   -99.26%  (p=0.000 n=10+10)
```

keegancsmith committed:
query: binary encoder for repo branches

We implement a custom binary marshaller for a list of repos to
branches. When profiling Sourcegraph this is one of the dominant items
and this provides a significant improvement.

This commit does not provide a way to roll this change out. I
investigated if this was possible with Gob and its pretty
difficult. This means that while Sourcegraph and Zoekt are out of sync
searches will fail. I am unsure if its worth making this smooth to
rollout.

  name                   old time/op    new time/op    delta
  RepoBranches_Encode-8    2.21ms ± 3%    0.63ms ± 5%  -71.55%  (p=0.000 n=10+10)
  RepoBranches_Decode-8    4.08ms ± 5%    1.20ms ± 1%  -70.58%  (p=0.000 n=10+10)

  name                   old bytes      new bytes      delta
  RepoBranches_Encode-8     393kB ± 0%     344kB ± 0%  -12.48%  (p=0.000 n=10+10)

  name                   old alloc/op   new alloc/op   delta
  RepoBranches_Encode-8     726kB ± 0%     344kB ± 0%  -52.59%  (p=0.000 n=6+9)
  RepoBranches_Decode-8    2.31MB ± 0%    1.58MB ± 0%  -31.62%  (p=0.000 n=10+10)

  name                   old allocs/op  new allocs/op  delta
  RepoBranches_Encode-8     20.0k ± 0%      0.0k ± 0%  -99.99%  (p=0.000 n=10+10)

Change-Id: I8f980b09c28e552935efa3c792f25c2aeec1e9fc

keegancsmith commented:
> Note: This PR does not provide a way to roll this change out. I
investigated if this was possible with Gob and its pretty
difficult. This means that while Sourcegraph and Zoekt are out of sync
searches will fail. I am unsure if its worth making this smooth to
rollout.

I can create a new type `RepoBranches2` and remove it once in two releases? Additionally we would need to add some sort of version to the API so the frontend would know whether it could use RepoBranches2 or the old version.

rvantonder reviewed (approved):
> I am unsure if its worth making this smooth to rollout. I can create a new type RepoBranches2 and remove it once in two releases?

I don't have a good idea here... it's your call :-)

bobheadxi reviewed (approved):
Pretty impressive gains 😮 

> This means that while Sourcegraph and Zoekt are out of sync searches will fail.

Does this happen frequently / for long enough to be problematic in practice?

keegancsmith commented:
> Does this happen frequently / for long enough to be problematic in practice?

StatefulSets roll out one replica at a time. So if you have 14 replicas
search would effectively be down for 5->10min?

I can engineer some proper distributed system rollout stuff here (not
the worst thing to do). Or for large scale instances the downtime could
be regarded as acceptable? For small scale instances there isn't as much
of an issue. @uwedeportivo what does your gut tell you? Can we have this
downtime on our next release.

keegancsmith committed:
correct comment on *Binary

Change-Id: I070c7be1af2ee4ed28b35e437e49e0f54c44cde2

keegancsmith committed:
only allocate HEAD list once

Change-Id: I63208eb14ad13d30915011f76a00716e2012e9e4

keegancsmith committed:
use shared buffer for strings when decoding

This avoids allocating a string per name. Instead we can take a copy of
the input bytes and create strings pointing into the byte slice.

In the benchmark this takes the number of allocations from ~20,000 to
~400!

Co-Author: tsenart

Change-Id: I916f3fc9083c90c091b8f03694bb48f9f0019b80

uwedeportivo commented:
>  @uwedeportivo what does your gut tell you? Can we have this downtime on our next release.

short answer: yes, go for it.

here's my thinking: this is a wire protocol, so no persistent state. our release process is as a monthly whole process, so all services get upgraded at the same time, so both sides of this wire exchange go out at the same time. a 10 min downtime (for bigdata installations) is acceptable in a month. this is weighing simplicity of code vs downtime and convenience of upgrading each service separately. the latter (almost) never happens for us and the former is no big deal for most installations.

i like that this wire protocol has a version so moving forward we can change it and support a smooth upgrade. for example if we change the len(branches) < 255 limit.

we should be fairly confident in this change (i think we are). rolling back in a later version makes the situation more confusing. not a biggie, but would be nice to avoid.

dot-com (cloud) would be the only case where i would worry about the rollout when this lands because of insiders images. not sure how important it is and how we should handle it.

keegancsmith commented:
Some hacking with @tsenart and we do a great trick were we avoid doing so many allocations when decoding. We use the classic unsafe pointer trick to make a string from a byte array. In the benchmark we went from doing 20k allocations to doing about 400! See [`0ecd1ac` (#53)](https://github.com/sourcegraph/zoekt/pull/53/commits/0ecd1acfdbb224e0da6fd03abad20329367207f7)

keegancsmith commented:
Great. I will roll this out tomorrow. I want to ensure I have reasonable way to validate the performance improvements. I'm very convinced it will help a great deal with garbage pressure which is harder to properly quantify the impact of.

I think I'll use a bit of a blackbox methodology. I'll run src-cli in a loop. After 1min I'll take a cpu profile of frontend. Then I'll run for 5min and see how many searches it managed to do in 5min.

Deploy this change and rinse repeat. Will do this against bigdata for the scale and no noise from other requests.

keegancsmith committed:
use constant for max size of a varint

Change-Id: Ibb674df6e94aa98c66958b4be08afbb93c408065

keegancsmith committed:
only allocate once in encode

Change-Id: Ie7c6fcc59b032d49dfae54f1f26dfeb3c5ac308a

keegancsmith committed:
less confusing comment about branches len

Change-Id: I47bb2cc31da0ddafcad1f30728dbd0df34ac2a9b

keegancsmith committed:
slightly clearer special casing of branches HEAD

Change-Id: I5115d21c5d5822cf419c319fb739747908648562

keegancsmith committed:
check for empty branches

Change-Id: I03a6bf7d1d76a67436850f1a19de49f6e5574ac2

keegancsmith committed:
update benchmark comment

Change-Id: Ib23e79ebcc4a62e30344418954c6d9ab561718a0
